File: apps/android/app/src/main/java/com/exponential/app/data/api/WorkspacesApi.kt
package com.exponential.app.data.api

import com.exponential.app.data.db.WorkspaceEntity
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.serialization.Serializable

@Serializable
data class EnsureDefaultResult(val workspace: WorkspaceEntity)

@Serializable
data class DeleteWorkspaceInput(val workspaceId: String)

@Serializable
data class DeleteProjectInput(val projectId: String)

@Serializable
data class WorkspaceBySlugInput(val slug: String)

/// Minimal projection returned by the public-aware `workspaces.getBySlug`
/// lookup. `hasPublicBoard` is true when the workspace hosts at least one
/// public feedback-board project; `membership` is the caller's role, null for
/// non-members (mirrors the web route guard).
@Serializable
data class WorkspaceBySlugResult(
    val id: String,
    val name: String,
    val slug: String,
    val iconUrl: String? = null,
    val hasPublicBoard: Boolean = false,
    val membership: String? = null,
)

@Serializable
private object EmptyInput

@Singleton
class WorkspacesApi @Inject constructor(private val trpc: TrpcClient) {

    suspend fun ensureDefault(accountId: String): WorkspaceEntity =
        trpc.mutation(
            accountId,
            path = "workspaces.ensureDefault",
            input = EmptyInput,
            inputSerializer = EmptyInput.serializer(),
            outputSerializer = EnsureDefaultResult.serializer(),
        ).workspace

    /// Public-aware workspace lookup by slug. Resolves workspaces the user has
    /// NOT synced yet (public boards only sync after an explicit join), so the
    /// app can offer the in-app join gate. Throws NOT_FOUND for private
    /// workspaces the caller can't access.
    suspend fun getBySlug(accountId: String, slug: String): WorkspaceBySlugResult =
        trpc.query(
            accountId,
            path = "workspaces.getBySlug",
            input = WorkspaceBySlugInput(slug),
            inputSerializer = WorkspaceBySlugInput.serializer(),
            outputSerializer = WorkspaceBySlugResult.serializer(),
        )

    suspend fun delete(accountId: String, workspaceId: String) {
        trpc.mutationUnit(
            accountId,
            path = "workspaces.delete",
            input = DeleteWorkspaceInput(workspaceId),
            inputSerializer = DeleteWorkspaceInput.serializer(),
        )
    }

    suspend fun deleteProject(accountId: String, projectId: String) {
        trpc.mutationUnit(
            accountId,
            path = "projects.delete",
            input = DeleteProjectInput(projectId),
            inputSerializer = DeleteProjectInput.serializer(),
        )
    }
}

File: apps/ios/Exponential/Data/API/WorkspacesApi.swift
import Foundation

struct EnsureDefaultResult: Decodable {
    let workspace: WorkspaceResult
}

struct WorkspaceResult: Decodable {
    let id: String
    let name: String
    let slug: String
}

struct UpdateWorkspaceInput: Encodable {
    let id: String
    var name: String?
    var isPublic: Bool?
    var publicWritePolicy: String?
    var iconUrl: String?
}

struct DeleteWorkspaceInput: Encodable {
    let workspaceId: String
}

struct DeleteProjectInput: Encodable {
    let projectId: String
}

private struct EmptyInput: Encodable {}
private struct EmptyResult: Decodable {}

final class WorkspacesApi: Sendable {
    private let trpc: TrpcClient

    init(trpc: TrpcClient) {
        self.trpc = trpc
    }

    func ensureDefault(accountId: String) async throws -> WorkspaceResult {
        let result: EnsureDefaultResult = try await trpc.mutation(accountId: accountId, path: "workspaces.ensureDefault", input: EmptyInput())
        return result.workspace
    }

    func update(accountId: String, _ input: UpdateWorkspaceInput) async throws {
        let _: EmptyResult = try await trpc.mutation(accountId: accountId, path: "workspaces.update", input: input)
    }

    func delete(accountId: String, workspaceId: String) async throws {
        try await trpc.mutationVoid(accountId: accountId, path: "workspaces.delete", input: DeleteWorkspaceInput(workspaceId: workspaceId))
    }

    func deleteProject(accountId: String, projectId: String) async throws {
        try await trpc.mutationVoid(accountId: accountId, path: "projects.delete", input: DeleteProjectInput(projectId: projectId))
    }
}
